ouroboros-network-protocols/src/Ouroboros/Network/Protocol/LocalStateQuery/Type.hs

```diff
@@ -212,3 +224,39 @@ instance (ShowQuery query, Show point)
         showsPrec 11 pt
       MsgDone ->
         showString "MsgDone"
+
+{-------------------------------------------------------------------------------
+  Footprints
+-------------------------------------------------------------------------------}
+
+-- | A singleton-like data family
+data family Sing (a :: k)
+
+-- | A singleton-like class
+class SingI (a :: k) where
+  sing :: Sing a
+
+-- | Queries on the local state might use ledger tables. This datatype (which
+-- will sometimes be concretized via @sing@) allows Consensus to categorize the
+-- queries.
+data QueryFootprint =
+    -- | The query doesn't need ledger tables, thus can be answered only with
+    -- the ledger state.
+    QFNone
+    -- | The query needs some tables, but doesn't need to traverse the whole
+    -- UTxO set.
+  | QFOne
+    -- | The query needs to traverse the whole UTxO set.
+  | QFAll
```

Comment:
I have an impression that this flag is query specific, isn't it?  It seems it only matters for the server side, not the client side.  If this is the case I'd prefer to keep in on the consensus side, because here we are complicating things also for clients, e.g. `cardano-wallet`, `cardano-cli`, etc.

If we move it to consensus, the queries still have the type and it will affect clients, so the question is if it is worth to encode this as a type parameter? Removing the type only make sense if each query is only done in one of the mode, rather than the client side choosing the mode of operation of a query.